### PR TITLE
i2c: fix issue of stm32-i2c-v2 (F0/F3/F7) driver hanging in endless loop

### LINKED ISSUE #4429: I2C: stm32-i2c-v2 Driver (F0/F3/F7) gets stuck in endless loop when handling restart conditions
in drivers/i2c/i2c_ll_stm32_v2.c:

When an i2c-message prepends a restart condition (I2C_MSG_RESTART flag is set) in function **msg_init** auto end mode is beeing disabled for this message, but function **msg_done** gets stuck in a loop forever, waiting for a stop condition on the bus.

Tested with stm32f3_disco board and samples/drivers/i2c_fujitsu_fram modified to use I2C_1

### PR DESCRIPTION
**Pull request by dwagenk**
Fixes #4429
Driver didn't work properly when a transfer consisted of multiple
messages.
Tested with stm32f3_disco board and samples/drivers/i2c_fujitsu_fram
example adapted to use I2C_1 as I2C_DEV

Signed-off-by: Daniel Wagenknecht <wagenknecht@clage.de>

### CODE REVIEW

File: drivers/i2c/i2c_ll_stm32_v2.c [hunk 1]
```diff
@@ -41,12 +41,16 @@ static inline void msg_init(struct device *dev, struct i2c_msg *msg,
 	LL_I2C_SetTransferRequest(i2c, transfer);
 	LL_I2C_SetTransferSize(i2c, len);
 
-	if ((flags & I2C_MSG_RESTART) == I2C_MSG_RESTART) {
+
+	/* Only activate auto end mode if this message doesn't prepend or follow a restart condition */
+	if (	((flags & I2C_MSG_RESTART) == I2C_MSG_RESTART)
+		||	(LL_I2C_IsActiveFlag_TC(i2c))	)	{
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32_v2.c]: Why do you check the transfer completed flag here? The transfer hasn't started yet.

File: drivers/i2c/i2c_ll_stm32_v2.c [hunk 2]
```diff
@@ -41,12 +41,16 @@ static inline void msg_init(struct device *dev, struct i2c_msg *msg,
...
 		LL_I2C_DisableAutoEndMode(i2c);
 	} else {
 		LL_I2C_EnableAutoEndMode(i2c);
 	}
 
+
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32_v2.c]: Remove this line

File: drivers/i2c/i2c_ll_stm32_v2.c [hunk 3]
```diff
@@ -56,11 +60,24 @@ static inline void msg_done(struct device *dev, unsigned int flags)
 	const struct i2c_stm32_config *cfg = DEV_CFG(dev);
 	I2C_TypeDef *i2c = cfg->i2c;
 
+	if (!LL_I2C_IsEnabledAutoEndMode(i2c))
+	{
+		/* Wait for transfer to complete */
+		while (!LL_I2C_IsActiveFlag_TC(i2c) ){
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32_v2.c]: Why wait only for transfer to complete when auto end mode is disabled and not always?

File: drivers/i2c/i2c_ll_stm32_v2.c [hunk 4]
```diff
@@ -56,11 +60,24 @@ static inline void msg_done(struct device *dev, unsigned int flags)
...
+			;
+		}
+		/* Issue stop condition if necessary */
+		if ((flags & I2C_MSG_RESTART) == 0) {
+			LL_I2C_GenerateStopCondition(i2c);
+		}
+	}
+
+	/* Wait for stop condition on bus if necessary */
 	if ((flags & I2C_MSG_RESTART) == 0) {
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32_v2.c]: The two if ((flags & I2C_MSG_RESTART) == 0) should be merged

**Review by ydamigos — CHANGES_REQUESTED** [review]
We still need to take care how the driver is handling the I2C messages. Probably in another PR.

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 1 above)
Comment (DWagenk) [reply drivers/i2c/i2c_ll_stm32_v2.c]: Yes thats right for the first message of a transfer: 
TC not set, RESTART set or not set, so auto end mode will only be disabled when RESTART is set.

But for messages following a non auto end message (previous message had RESTART set):
TC is set -> This message needs to be ended manually. If we would activate  auto end mode here, it would lead to an immediate Stop Condition (because previous message is done, but waiting for a stop or restart condition)

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 3 above)
Comment (DWagenk) [reply drivers/i2c/i2c_ll_stm32_v2.c]: If auto end mode is activated TC flag will not be set.
Reference manual (RM0316) for F3 says: 
> TC: Transfer Complete (master mode)
This flag is set by hardware when RELOAD=0, AUTOEND=0 and NBYTES data have been
transferred. It is cleared by software when START bit or STOP bit is set.
Note: This bit is cleared by hardware when PE=0.

**So when auto end mode is activated the msg_done function will read like this:**
76: RESTART = 0, so if evaluates to true
77: Wait for stop condition appearing on bus
80: Clear stop flag
done

**With auto end mode disabled:**
63: if evaluates to true
66: wait for transfer complete flag

now if RESTART = 0 (this message should be followed by a stop condition)
70: if evaluates to true
71: stop condition is generated manually
76: if evaluates to true
77: Wait for stop condition appearing on bus

but if RESTART = 1 (don't generate a stop condition)
70: if evaluates to false
76: if evaluates to false
done

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 4 above)
Comment (DWagenk) [reply drivers/i2c/i2c_ll_stm32_v2.c]: well, this last if ((flags & I2C_MSG_RESTART) == 0) needs to be evaluated for both auto end and none auto end mode. So I don't see how they can be merged without breaking functionality or needing two if(!LL_I2C_IsEnabledAutoEndMode)

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 3 above)
Comment (ydamigos) [reply drivers/i2c/i2c_ll_stm32_v2.c]: > If auto end mode is activated TC flag will not be set.

You are right. I missed that.

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 1 above)
Comment (ydamigos) [reply drivers/i2c/i2c_ll_stm32_v2.c]: >But for messages following a non auto end message (previous message had RESTART set):
TC is set -> This message needs to be ended manually. If we would activate auto end mode here, it would lead to an immediate Stop Condition (because previous message is done, but waiting for a stop or restart condition)

Even if we don't enable auto end mode, an early stop condition may be generated in msg_done, as the TC flag is set and the while will evaluate to false. We could clear it inside the if but it turns out to be an ugly hack and make hard to follow the code. I believe we should not use auto end mode and handle it in software.

Comment (DWagenk) [reply drivers/i2c/i2c_ll_stm32_v2.c]: I'd be OK with not using auto end mode. Certainly will make the Code easier to read.

**Comment by DWagenk** [comment]
@ydamigos 
I updated the commit to not use auto end mode anymore. Code is really more straightforward this way.

Sorry for the many updates to the branch, I didn't have checkpatch, gitlint and sanitycheck set up. I'll have a deeper look into that before my next contributions.

File: drivers/i2c/i2c_ll_stm32_v2.c [hunk 5]
```diff
@@ -248,7 +250,7 @@ int stm32_i2c_msg_read(struct device *dev, struct i2c_msg *msg,
 		len--;
 	}
 
-	msg_done(dev, flags | msg->flags);
+	msg_done(dev, flags);
```
[the remark this answers is not in the document]
Comment (adilen) [reply drivers/i2c/i2c_ll_stm32_v2.c]: The flag I2C_MSG_RESTART inside this driver means 
>this message is followed by a restart condition, so don't issue a stop condition at the end of this message.

This differs from the meaning of the I2C_MSG_RESTART in the general Zephyr I2C API
>I2C_MSG_RESTART (1 << 2)
RESTART I2C transaction for this message.

To fix this both the stm32-i2c-v1 and v2 driver need a rework, which is, why I opened Issue #4459.
 
the variable flags is set to I2C_MSG_RESTART in i2c_ll_stm32.c if the current message is followed by one that has a different direction (R/W). So passing it into msg_done function leads to no stop condition beeing issued between those two messages, like intended.

msg->flags is controlled by the app/user, having the API meaning of the I2C_MSG_RESTART flag in mind. Passing it into msg_done would lead to no stop condition being issued after this message, while the user/app intended to start this message with a restart condition.

So until the meanings of I2C_MSG_RESTART in stm32 drivers and API are harmonized we should not use msg->flags inside this driver. 

Daniel Wagenknecht

**Review by ydamigos — APPROVED** [review]

**Comment by ydamigos** [comment]
Please rebase the PR against latest arm branch.

File: drivers/i2c/i2c_ll_stm32_v2.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (DWagenk) [reply drivers/i2c/i2c_ll_stm32_v2.c]: just saw, that this applys to the interrrupt driven driver as well (line 182/184). Changed it with the latest update of this pull request (and rebased onto latest arm branch).

**Comment by DWagenk** [comment]
@galak thanks for updating the arm branch, that solved it...

**Comment by galak** [comment]
@ydamigos let me know if this now looks good to merge

**Comment by ydamigos** [comment]
@galak LGTM

**Review by galak — APPROVED** [review]

**Comment by ydamigos** [comment]
@galak I cannot find this commit in arm branch. Maybe something went wrong after a rebase.

**Comment by galak** [comment]
@ydamigos thanks for catching that, guessing it got lost in a rebase.  Should be fixed now.
